### PR TITLE
feat: add basic pipeline cleanup actor option

### PR DESCRIPTION
**Pull request by JonathanGriffe**

### CODE REVIEW

File: remoulade/composition.py [hunk 1]
```diff
@@ -27,6 +27,8 @@
     from .message import Message  # noqa
     from .result import Result
 
+Actor = TypeVar("Actor")
```
Comment (antoinerabany) [inline remoulade/composition.py:30]: why use this, you should use the Actor in actor.py

File: remoulade/composition.py [hunk 2]
```diff
@@ -143,6 +151,9 @@ def run(self, *, delay: Optional[int] = None) -> "pipeline":
             self.broker.enqueue(first, delay=delay)
         return self
 
+    def add_cleanup(self, cleanup_actor: Union[str, Actor]):
```
Comment (antoinerabany) [inline remoulade/composition.py:154]: why you need this ? cleanup_actor can be accessed

File: remoulade/middleware/retries.py [hunk 3]
```diff
@@ -86,8 +86,9 @@ def after_process_message(self, broker, message, *, result=None, exception=None)
             message.fail()
             return
 
-        message.options["retries"] += 1
-        message.options["traceback"] = traceback.format_exc(limit=30)
+        new_message = message.copy()
+        new_message.options["retries"] += 1
```
Comment (antoinerabany) [inline remoulade/middleware/retries.py:90]: we need a test of this beahvior

**Review by antoinerabany — CHANGES_REQUESTED** [review]

File: remoulade/middleware/retries.py (hunk 3 above)
Comment (JonathanGriffe) [reply remoulade/middleware/retries.py:90]: This part is no longer necessary, do we still keep it ?

File: remoulade/composition.py [hunk 4]
```diff
@@ -65,8 +66,11 @@ class pipeline:
         children(List[Message|group]) The sequence of messages or groups to execute as a pipeline
     """
 
-    def __init__(self, children: "Iterable[Union[Message, pipeline, group]]"):
+    def __init__(
+        self, children: "Iterable[Union[Message, pipeline, group]]", cleanup_actor: "Optional[Union[str, Actor]]" = None
```
Comment (antoinerabany) [inline remoulade/composition.py:70]: remove from composotion ?

File: remoulade/middleware/catch_error.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
+from remoulade import Middleware
+from remoulade.middleware import Retries
+
+
+class CatchError(Middleware):
```
Comment (antoinerabany) [inline remoulade/middleware/catch_error.py:5]: test this middleware

File: tests/middleware/test_retries.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    retries_original_message, retries_retried_message = None, None
+
+    class TestMiddleware(Middleware):
+
+        default_before = Retries
+
+        def after_process_message(self, broker, message, *, result=None, exception=None):
+            nonlocal retries_original_message, retries_retried_message
+            if retries_original_message is None:
+                print(message.options)
```
Comment (antoinerabany) [inline tests/middleware/test_retries.py:16]: print ?

**Review by antoinerabany — CHANGES_REQUESTED** [review]

File: tests/middleware/test_catch_error.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    cleanup_count = 0
+
+    @remoulade.actor
+    def cleanup(actor_name, exception_name, args, kwargs):
+        nonlocal cleanup_count
+        cleanup_count += 1
+
+    @remoulade.actor
+    def fail_actor():
+        raise Exception
```
Comment (antoinerabany) [inline tests/middleware/test_catch_error.py:17]: raise Exception()

**Review by antoinerabany — APPROVED** [review]

**Pull request merged**